crates/holochain/src/conductor/conductor.rs

```diff
@@ -128,26 +132,26 @@ where
 
     /// The caches databases. These are shared across cells.
     /// There is one per unique Dna.
-    caches: parking_lot::Mutex<HashMap<DnaHash, EnvWrite>>,
+    caches: RwShare<HashMap<DnaHash, EnvWrite>>,
 
     /// Set to true when `conductor.shutdown()` has been called, so that other
     /// tasks can check on the shutdown status
-    shutting_down: bool,
+    shutting_down: Arc<AtomicBool>,
 
     /// The admin websocket ports this conductor has open.
     /// This exists so that we can run tests and bind to port 0, and find out
     /// the dynamically allocated port later.
-    admin_websocket_ports: Vec<u16>,
+    admin_websocket_ports: RwShare<Vec<u16>>,
 
     /// Collection app interface data, keyed by id
-    app_interfaces: HashMap<AppInterfaceId, AppInterfaceRuntime>,
+    app_interfaces: RwShare<HashMap<AppInterfaceId, AppInterfaceRuntime>>,
 
     /// The channels and handles needed to interact with the task_manager task.
     /// If this is None, then the task manager has not yet been initialized.
-    pub(super) task_manager: Option<TaskManagerClient>,
+    pub(super) task_manager: RwShare<Option<TaskManagerClient>>,
 
     /// Placeholder for what will be the real DNA/Wasm cache
-    dna_store: DS,
+    dna_store: RwShare<DS>,
```

Comment:
Not sure how good of an idea it is to have all of this substate independently lockable. There's more opportunity for race conditions. For instance, what if we add a Cell but somehow didn't get around to updating the DnaStore yet? Or vice versa during uninstallation? Should some of these things be locked together?

Reply:
I hear what you're saying but I'm not sure it's worth locking all the conductor to enforce this. It makes it much harder to work with because you have to pull all the data out before doing any async work which in some cases might be impossible. In your example you don't actually need an atomic transaction, you just need the dna lock to happen before the cell lock. If you really do need them to happen together you can take one lock inside the other but I feel this should be opt into rather then the default.
If you feel like there's some data that really should be coupled then we could look at merging a few of these but from what I saw when I was refactoring this yesterday we were often just taking the same lock out over and over because they can't be held for long periods anyway.

Reply:
Yeah that's fair...guess we're at the point where we have to care more about performance, meaning we have to keep more things in mind while we write